File: BOJ/28_트리/2263_트리의순회.py
import sys
sys.setrecursionlimit(int(1e6))
read = sys.stdin.readline

def preorder(in_L,in_R,po_L,po_R) :

    if (in_L > in_R) or (po_L > po_R) : return

    root = postorder[po_R]
    # 이거 쓰면 시간 초과
    # L = inorder.index(root) - in_L
    L = inIdx[root] - in_L
    R = N - 1 - L

    print(root,end=' ')

    preorder(in_L, in_L+L-1, po_L, po_L+L-1)
    preorder(in_L+L+1, in_R, po_L+L, po_R-1)


N = int(read())
inorder = list(map(int,read().split()))
postorder = list(map(int, read().split()))

inIdx = [0 for _ in range(N+1)]
for i in range(N) :
    inIdx[inorder[i]] = i

preorder(0,N-1,0,N-1)

File: BOJ/28_트리/2263_트리의순회.cpp
#include <iostream>

#define MAXN 100000

using namespace std;
int inorder[MAXN+1],postorder[MAXN+1],inIdx[MAXN+1];
int N;

void preorder(int in_L, int in_R, int po_L, int po_R) {

    if((in_L > in_R) || (po_L > po_R)) return ;

    int root = postorder[po_R];
    int L = inIdx[root] - in_L;

    printf("%d ",root);
    preorder(in_L,in_L+L-1,po_L,po_L+L-1);
    preorder(in_L+L+1,in_R,po_L+L,po_R-1);

}

int main() {

    scanf("%d",&N);
    for(int i=0; i<N; i++) scanf("%d",inorder+i);
    for(int i=0; i<N; i++) scanf("%d",postorder+i);
    for(int i=0; i<N; i++) inIdx[inorder[i]] = i;
    
    preorder(0,N-1,0,N-1);

    return 0;
}